### PR TITLE
LLT-4858: Implement nat state reporting

### PR DESCRIPTION
**Pull request by tomaszklak**
### Problem
We need to report the nat analytics as described in RFC LLT-0051 and RFC LLT-0058.

### Solution
High level list of changes:
- relayed connection state changes are 'attached' to the node events reporting - this should cover appearing and disappearing of new nodes (which start in relayed state)
- changes between direct and relayed are happening in 3 places. This is mostly because which peer is reporting the analytics (we chose peer with smaller key) can differ from which peer initiated the upgrade:
  - event for the beginning of direct connection can happen in one of the 2 places:
    - when the peer with the winning key has sent the upgrade request and received response with decision to accept the upgrade
    - when the peer with the winning key has received the upgrade request and has accepted it
  - when the wg_controller has detected a downgrade

I've also added nat-lab test based on the logs to have some sort of sanity check - the actual analytics reporting is being done in parallel. When it's done that test can be updated and other might be added.

### :ballot_box_with_check: Definition of Done checklist
- [x] Commit history is clean ([requirements](../blob/main/docs/git_commit_messages_requirements.md))
- [x] README.md is updated
- [x] changelog.md is updated
- [x] Functionality is covered by unit or integration tests

### CODE REVIEW

**Review by gytsto — COMMENTED** [review]
+0.5

File: crates/telio-traversal/src/upgrade_sync.rs [hunk 1]
```diff
@@ -43,6 +51,20 @@ pub struct UpgradeRequest {
     pub session: Session,
 }
 
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+enum Direction<T: Debug + PartialEq + Eq + Hash + Clone + Copy> {
+    Sent(T),
+    Received(T),
+}
```
Comment (stalowyjez) [inline crates/telio-traversal/src/upgrade_sync.rs:58]: Maybe `Directed`? Maybe you could add some comment why do we need it?

File: crates/telio-traversal/src/upgrade_sync.rs [hunk 2]
```diff
@@ -53,32 +75,49 @@ pub trait UpgradeSyncTrait {
...
+    upgrade_requests: HashMap<Direction<PublicKey>, UpgradeRequest>,
     expiration_period: Duration,
     upgrade_controller: Arc<dyn UpgradeController>,
     poll_timer: Interval,
     upgrade_decision_intercoms: Chan<(PublicKey, UpgradeDecisionMsg)>,
+    // Collection of not yet confirmed Upgrade requests
+    pending_direct_sessions: LruCache<Session, PendingSessionData>,
+    our_public_key: PublicKey,
+    connectivity_data_aggregator: Arc<ConnectivityDataAggregator>,
+    wireguard: Arc<dyn WireGuard>,
```
Comment (stalowyjez) [inline crates/telio-traversal/src/upgrade_sync.rs:107]: Well, I'm not sure if we want to use wireguard interface here - if we need to create AnalyticsEvent here, maybe it would be better to change it to some simpler struct and gather this data in aggregator, where we already have this data?

File: crates/telio-traversal/src/upgrade_sync.rs [hunk 3]
```diff
@@ -195,23 +308,70 @@ impl State {
...
+                        {
+                            // We reject this request, because we sent our own which is not
+                            // yet expired and our key is winning - the other side should accept
+                            // our request instead.
+                            telio_proto::Decision::RejectedDueToConcurrentUpgrade
+                        } else {
+                            telio_proto::Decision::Accepted
+                        }
+                    }
+                    None => telio_proto::Decision::Accepted,
```
Comment (stalowyjez) [inline crates/telio-traversal/src/upgrade_sync.rs:326]: I feel like this should be in different PR tbh

File: crates/telio-crypto/src/lib.rs [hunk 4]
```diff
@@ -42,6 +42,16 @@ pub struct SecretKey([u8; KEY_SIZE]);
 )]
 pub struct PublicKey(pub [u8; KEY_SIZE]);
 
+/// Canonical way to order keys
+pub fn key_order(lhs: &PublicKey, rhs: &PublicKey) -> Ordering {
+    lhs.cmp(rhs)
+}
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-crypto/src/lib.rs:48]: I fell that such an explicit function is better to indicate significance of this particular order. The fact that this is the 'natural' order for this type is just a coincidence - had the heartbeat algorithm used the biggest key, this would be the reverse order.

File: crates/telio-proto/protos/upgrade.proto [hunk 5]
```diff
@@ -9,6 +9,10 @@ enum Decision {
 message Upgrade {
 	string endpoint = 1;
 	fixed64 session = 2;
+	// endpoint type used by the sender
+	uint32 endpoint_type = 3;
+	// endpoint type selected for the receiver
+	uint32 receiver_endpoint_type = 4;
 }
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-proto/protos/upgrade.proto:16]: The pong message tells the upgrade initiator the type of the endpoint of the other side. Here we are telling the upgrade responder the type of our endpoint, and the type of theirs endpoint. You could argue that the second type is not strictly needed but I disagree. Since this is a decision made by the initiator, I think that the responder shouldn't guess from the endpoint its type (even if it is technically possible).

File: crates/telio-nurse/src/aggregator.rs [hunk 6]
```diff
@@ -430,30 +529,104 @@ mod tests {
...
+        connectivity_data_aggregator: ConnectivityDataAggregator,
+        losing_key: PublicKey,
+        winning_key: PublicKey,
+    }
+
+    #[derive(PartialEq, Eq, PartialOrd, Ord, Clone, Copy)]
+    enum KeyKind {
+        Winning,
+        Losing,
+    }
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-nurse/src/aggregator.rs:542]: I think that the name is fine, do you have any proposal what would be a better names here?

File: crates/telio-traversal/src/endpoint_providers/mod.rs [hunk 7]
```diff
@@ -90,6 +90,16 @@ pub enum EndpointProviderType {
...
+impl From<EndpointProviderType> for telio_model::features::EndpointProvider {
+    fn from(value: EndpointProviderType) -> Self {
+        match value {
+            EndpointProviderType::LocalInterfaces => telio_model::features::EndpointProvider::Local,
+            EndpointProviderType::Stun => telio_model::features::EndpointProvider::Stun,
+            EndpointProviderType::Upnp => telio_model::features::EndpointProvider::Upnp,
+        }
+    }
+}
+
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-traversal/src/endpoint_providers/mod.rs:102]: Not sure, but I didn't want to make this pr even bigger! 😓

File: crates/telio-wg/src/uapi.rs [hunk 8]
```diff
@@ -401,6 +401,45 @@ impl Peer {
...
+            }
+        }
+        if target != (None, None) {
+            dual_ip_addresses.push(match DualTarget::new(target) {
+                Ok(dt) => dt,
+                Err(DualTargetError::NoTarget) => DualTarget::default(),
+            });
+        }
+        dual_ip_addresses
+    }
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-wg/src/uapi.rs:442]: This is not a new code, it was moved from `crates/telio-wg/src/wg.rs`. You might be right, I didn't look into it, just extracted it to a function.

File: src/device/mod.rs [hunk 9]
```diff
@@ -2027,6 +2045,19 @@ impl TaskRuntime for Runtime {
...
                 if let Some(node) = node {
+                    if mesh_event.state != PeerState::Connecting && node.path == PathType::Relay {
+                        let event = AnalyticsEvent{
+                            public_key: mesh_event.peer.public_key,
+                            dual_ip_addresses: mesh_event.peer.get_dual_ip_addresses(),
+                            tx_bytes: mesh_event.peer.tx_bytes.unwrap_or_default(),
+                            rx_bytes: mesh_event.peer.rx_bytes.unwrap_or_default(),
+                            peer_state: mesh_event.state,
+                            timestamp: Instant::now(),
+                        };
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply src/device/mod.rs:2056]: The only difference the constructor would make is that it could automatically fill `timestamp`. Not sure that this is worth it. Let me know what you think.

File: nat-lab/tests/test_direct_connection.py [hunk 10]
```diff
@@ -270,17 +279,36 @@ async def test_direct_working_paths(
             await testing.wait_long(ping.wait_for_next_ping())
 
 
+@pytest.mark.moose
 @pytest.mark.asyncio
 @pytest.mark.timeout(240)
 @pytest.mark.parametrize("setup_params, reflexive_ip", UHP_WORKING_PATHS)
-async def test_direct_working_paths_are_reestablished(
+async def test_direct_working_paths_are_reestablished_and_correctly_reported_in_analytics(
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply nat-lab/tests/test_direct_connection.py:286]: Not sure how would you wan to split it. Most of the tests are already being added in #396 - here I want to have a bigger sanity check that even when we have direct/relay toggling we will still report proper sequence of events.

Let me know what you had in mind here.

File: crates/telio-traversal/src/upgrade_sync.rs (hunk 2 above)
Comment (tomaszklak) [reply crates/telio-traversal/src/upgrade_sync.rs:107]: I'm not convinced. If I have to be honest, I think that its aggregator that should not know about wireguard, not UpgradeSync.

File: crates/telio-crypto/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (stalowyjez) [reply crates/telio-crypto/src/lib.rs:48]: In such case it probably doesn't need to be an order, as we don't need transitivity here, but it might be just any full antisymmetric relation, i.e. we just need to know which node to pick, so imho returning a boolean, or maybe this `Winning` / `Losing` enum would be clearer? I think also some more verbose comment would be helpful, maybe with some explanation, why we need this functions and the RFC number? I don't think that "canonical" helps a lot here.

Also, I feel like all of this logic with picking the right node doesn't really touch the NAT state reporting, so maybe it could be moved to a separate PR?

File: src/device/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (stalowyjez) [reply src/device/mod.rs:2056]: I think that moving this logic to some `From` trait and just using it here would be nice

File: crates/telio-crypto/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (stalowyjez) [reply crates/telio-crypto/src/lib.rs:48]: Oh, there is even a proper name for this one: https://en.wikipedia.org/wiki/Tournament_(graph_theory)

[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-crypto/src/lib.rs:48]: Nat state reporting is being done *only* by the winning node so I'm not sure how to move it to a different PR.

File: crates/telio-crypto/src/lib.rs [hunk 11]
```diff
@@ -42,6 +42,16 @@ pub struct SecretKey([u8; KEY_SIZE]);
...
+
+/// Find a winning key using canonical ordering
+pub fn winning_key<'a>(keys: impl IntoIterator<Item = &'a PublicKey>) -> Option<&'a PublicKey> {
+    keys.into_iter().min_by(|l, r| key_order(l, r))
+}
+
```
[the remark this answers is not in the document]
Comment (stalowyjez) [reply crates/telio-crypto/src/lib.rs:54]: Ok, there are some problems I didn't see before:
 1. The purpose of this function is to compare between two nodes, this and remote - there is no use in considering more or less keys, we only lose one compile time check with this.
 2. It creates an unnecessary option just for the case of the empty list (which should not be the case, but if someone would make such mistake it is not event handled in any way e.g. `upgrade_sync.rs:316`)
 2. Maybe we would like to handle somehow case in which the keys are the same - it shouldn't happen and iterator just returns the first one in such case (but maybe that's fine?)

[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-crypto/src/lib.rs:54]: Why `max` instead? The rfc talks specifically about smallest key, so that we are consistent with how we resolve the meshned_id for heartbeat.

File: crates/telio-traversal/src/upgrade_sync.rs (hunk 2 above)
Comment (stalowyjez) [reply crates/telio-traversal/src/upgrade_sync.rs:107]: It might feel this way, but aggregator has to have it anyway and here it is unnecessary when it already has it

File: crates/telio-crypto/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (stalowyjez) [reply crates/telio-crypto/src/lib.rs:48]: Well, "NAT state reporting" depends on "winning and losing nodes", but we can have "losing and winning nodes" merged without "NAT state reporting" with no harm?

File: src/device/mod.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (tomaszklak) [reply src/device/mod.rs:2056]: I don't think this is good idea - this is not a lossless conversion. See: https://doc.rust-lang.org/std/convert/trait.From.html#when-to-implement-from

File: crates/telio-traversal/src/upgrade_sync.rs (hunk 1 above)
Comment (tomaszklak) [reply crates/telio-traversal/src/upgrade_sync.rs:58]: I'm not sure about the name change - I feel that types should generally be nouns, not verbs. But for sure a comment is needed here! :)

File: crates/telio-crypto/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-crypto/src/lib.rs:48]: I see what you mean. Do you think it has to be in a separate PR or would it be enough to extract this to a separate (first) commit in this pr?

[the remark this answers is not in the document]
Comment (stalowyjez) [reply crates/telio-crypto/src/lib.rs:48]: A separate commit sounds like a great improvement, but I think a separate PR would be even better (and relatively simple).

File: crates/telio-traversal/src/upgrade_sync.rs (hunk 2 above)
Comment (tomaszklak) [reply crates/telio-traversal/src/upgrade_sync.rs:109]: I don't think that the argument that Aggregator is already contaminated is an argument for putting in it even more wireguard. Since Aggregator is the first thing that depends on Wireguard interface in the telio-nurse, I think we should hope to remove that dependency in future.

File: crates/telio-crypto/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-crypto/src/lib.rs]: I think it's better to keep this in the same pr. If I move the key_order related changes to a separate PR it will look strange since there will be no other place where it is used and I think such refactoring should be done only when you have a second call site that will use the newly extracted function.

File: crates/telio-traversal/src/upgrade_sync.rs (hunk 3 above)
Comment (tomaszklak) [reply crates/telio-traversal/src/upgrade_sync.rs:328]: That's an integral part of the rfc 😕 Why do you think this shouldn't be a part of this PR?

File: crates/telio-nurse/src/aggregator.rs [hunk 12]
```diff
@@ -250,12 +254,12 @@ impl ConnectivityDataAggregator {
...
     /// * `event` - Analytic event with the current relayed peer state.
-    pub(crate) async fn change_peer_state_relayed(&self, event: &AnalyticsEvent) {
+    pub async fn change_peer_state_relayed(&self, event: &AnalyticsEvent) {
         self.change_peer_state_common(
             event,
             PeerEndpointTypes {
-                initiator_ep: EndpointType::Relay,
-                responder_ep: EndpointType::Relay,
+                local_ep: EndpointType::Relay,
+                remote_ep: EndpointType::Relay,
```
Comment (stalowyjez) [inline crates/telio-nurse/src/aggregator.rs:262]: Move this to separate commit

File: crates/telio-nurse/src/aggregator.rs [hunk 13]
```diff
@@ -444,30 +488,104 @@ mod tests {
...
+                        },
+                    )]
+                    .iter()
+                    .cloned()
+                    .collect();
+                    (keys[0].1, i, keys[2].1)
+                }
+                None => (keys[0].1, i, keys[2].1),
+            }
+        };
```
Comment (stalowyjez) [inline crates/telio-nurse/src/aggregator.rs:562]: It owuld be nice to extract this test refactor to a separate commit

File: crates/telio-traversal/src/upgrade_sync.rs (hunk 3 above)
Comment (stalowyjez) [reply crates/telio-traversal/src/upgrade_sync.rs:328]: You can move this to `winning_key` function commit

File: crates/telio-traversal/src/upgrade_sync.rs [hunk 14]
```diff
@@ -158,23 +241,53 @@ impl State {
...
+                public_key,
+                dual_ip_addresses: peer.get_dual_ip_addresses(),
+                tx_bytes: peer.tx_bytes.unwrap_or_default(),
+                rx_bytes: peer.rx_bytes.unwrap_or_default(),
+                peer_state: PeerState::Connected,
+                timestamp: Instant::now().into_std(),
+            };
+            self.connectivity_data_aggregator
+                .change_peer_state_relayed(&event)
+                .await;
```
Comment (stalowyjez) [inline crates/telio-traversal/src/upgrade_sync.rs:289]: Maybe move this to separate function

File: crates/telio-wg/src/uapi.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-wg/src/uapi.rs:458]: I've added a comment explaining the behaviour and also a test to verify it.

File: crates/telio-crypto/src/lib.rs [hunk 15]
```diff
@@ -42,6 +42,20 @@ pub struct SecretKey([u8; KEY_SIZE]);
...
 pub struct PublicKey(pub [u8; KEY_SIZE]);
 
+/// Canonical way to order keys as defined in RFC LLT-0051
+pub fn meshnet_canonical_key_order(lhs: &PublicKey, rhs: &PublicKey) -> Ordering {
+    lhs.cmp(rhs)
+}
+
+/// Find a winning key using canonical ordering from RFC LLT-0051
+pub fn winning_key<'a>(lhs: &'a PublicKey, rhs: &'a PublicKey) -> &'a PublicKey {
+    if lhs <= rhs {
```
[the remark this answers is not in the document]
Comment (stalowyjez) [reply crates/telio-crypto/src/lib.rs:52]: Also, what about renaming it to be consistent with the second function?

File: src/device/mod.rs [hunk 16]
```diff
@@ -2037,6 +2053,8 @@ impl TaskRuntime for Runtime {
                                     .await?;
 
                                 direct_entities.session_keeper.remove_node(&public_key).await?;
+
+                                direct_entities.upgrade_sync.downgrade(&public_key).await?;
```
Comment (stalowyjez) [inline src/device/mod.rs:2057]: It it necessary to use `upgrade_sync` for that? Couldn't you reuse `AnalyticsEvent::from_event` also here?

**Review by stalowyjez — COMMENTED** [review]
Looks quite good now, just a few minor comments

File: crates/telio-crypto/src/lib.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-crypto/src/lib.rs:52]: Duh! I have no idea why I used the `<=`... will fix that in a sec. As far as the name - what do you have in mind? I'm not sure how to make it consistent with the other one.

File: crates/telio-traversal/src/upgrade_sync.rs [hunk 17]
```diff
@@ -43,6 +45,22 @@ pub struct UpgradeRequest {
     pub session: Session,
 }
 
+/// Used to indicate if the inner value was received from the other peer, or
+/// sent to it.
+#[derive(Debug, PartialEq, Eq, Hash, Clone, Copy)]
+enum Direction<T: Debug + PartialEq + Eq + Hash + Clone + Copy> {
+    Sent(T),
+    Received(T),
+}
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-traversal/src/upgrade_sync.rs:60]: That's not really possible - it is used as a wrapper for `PublicKey` in a `HashMap` key.

File: crates/telio-traversal/src/upgrade_sync.rs [hunk 18]
```diff
@@ -301,10 +363,20 @@ impl State {
         }
 
         // Remove all expired events from our state
-        self.upgrade_requests.retain(|_, v| !is_expired(v));
+        let expiration_period = self.expiration_period;
+        self.upgrade_requests
+            .retain(|_, v| !Self::is_expired(expiration_period, v));
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-traversal/src/upgrade_sync.rs:441]: ratin borrows mutably so, it's not possible to have immutable borrow in the passed lambda.

File: crates/telio-traversal/src/upgrade_sync.rs [hunk 19]
```diff
@@ -325,17 +397,17 @@ impl Runtime for State {
...
                     Decision::Accepted => {},
                     Decision::RejectedDueToUnknownSession => {
                         if let Some(pk) = self.upgrade_requests.iter().find(|(_, req)| req.session == msg.session).map(|(pk,_)| *pk) {
-                            let _ = self.upgrade_controller.notify_failed_wg_connection(pk).await;
+                            let _ = self.upgrade_controller.notify_failed_wg_connection(pk.into()).await;
                         }
                     }
                     Decision::RejectedDueToConcurrentUpgrade => {
-                        // TODO: in LLT-4858
+                        self.upgrade_requests.retain(|pk, req| !(pk == &Direction::Sent(public_key) && req.session == msg.session));
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-traversal/src/upgrade_sync.rs:516]: Both are actually the same kind - remote.

**Review by stalowyjez — COMMENTED** [review]
Looks good to me, +1

File: crates/telio-nurse/src/qos.rs [hunk 20]
```diff
@@ -360,7 +361,7 @@ impl Analytics {
     ///
     /// * `event` - Received WG event.
     async fn handle_wg_event(&mut self, event: &AnalyticsEvent) {
-        if event.is_from_virtual_peer() {
+        if is_from_ignored_peer(event) {
             telio_log_trace!("WG event (Virtual peer, skipped): {:?}", event);
```
Comment (gindoslaw) [inline crates/telio-nurse/src/qos.rs:365]: Function "ignore peer"
Log message  "Virtual peer"

Comment (tomaszklak) [reply crates/telio-nurse/src/qos.rs]: Hope it's better now 😄

File: crates/telio-nurse/src/qos.rs [hunk 21]
```diff
@@ -519,6 +520,26 @@ impl Analytics {
     }
 }
 
+fn is_from_ignored_peer(event: &AnalyticsEvent) -> bool {
```
[the remark this answers is not in the document]
Comment (packgron) [reply crates/telio-nurse/src/qos.rs:523]: Ok it's kindof used in other areas, but still feels it would be nice to move it to be a member funtion of anlytics event

File: crates/telio-nurse/src/heartbeat.rs [hunk 22]
```diff
@@ -186,7 +186,7 @@ impl Collection {
         self.meshnet_ids
             .iter()
             .filter(|(_, id)| maximal_ids.contains(*id))
-            .min_by(|l, r| l.0.cmp(r.0))
+            .min_by(|l, r| meshnet_canonical_key_order(l.0, r.0))
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-nurse/src/heartbeat.rs:189]: Yeah, it was supposed to be named for documentation purposes - it's defined in rfc and we want it to be the same as the order used in heartbeat to select meshnet ids. But you are the second person that suggests this is useless so maybe I should just remove it and use the `PartialOrd` 🤔

File: crates/telio-nurse/src/qos.rs [hunk 23]
```diff
@@ -519,6 +520,26 @@ impl Analytics {
...
+    // The only allowed virtual peer for analytics is the VPN peer.
+    fn is_vpn(ip: IpAddr) -> bool {
+        [
+            IpAddr::V4(Ipv4Addr::new(100, 64, 0, 1)),
```
[the remark this answers is not in the document]
Comment (tomaszklak) [reply crates/telio-nurse/src/qos.rs:527]: You are right, I wasn't aware!

Sadly, I think it's broken - correct me if I'm wrong but the ipv6 address of vpn in `constants.rs` seems to be `fd74:fd74:696f::1` but it should be `fd74:656c:696f::1`.

File: nat-lab/tests/telio.py [hunk 24]
```diff
@@ -774,12 +773,17 @@ def get_stdout(self) -> str:
...
 
+    def get_features(self) -> TelioFeatures:
+        assert self._telio_features
+        return self._telio_features
+
     async def stop_device(self, timeout: float = 5) -> None:
         await asyncio.wait_for(self._write_command(["dev", "stop"]), timeout)
         self._interface_configured = False
-        assert Counter(self.get_runtime().get_started_tasks()) == Counter(
+        diff = set(self.get_runtime().get_started_tasks()) - set(
```
Comment (gytsto) [inline nat-lab/tests/telio.py:783]: Counter and set works differently here. Counter was chosen for a reason, to catch multiple occurrences of same task starting or stopping. For example, if 1 task starts 2 times, `set` will basically skip it, but `Counter` not.

```
from collections import Counter

start_task_list = ["task1", "task1", "task2"]
stop_task_list = ["task1", "task2"]

print(set(start_task_list) == set(stop_task_list))  # True
print(Counter(start_task_list) == Counter(stop_task_list))  # False

print(set(start_task_list) - set(stop_task_list))  # set()
print(Counter(start_task_list) - Counter(stop_task_list))  # Counter({'task1': 1})
```

Comment (gytsto) [reply nat-lab/tests/telio.py:783]: Would be nice to leave assert logic, so we could have `diff` printed in logs, but `Counter` should be brought back imo

File: crates/telio-nurse/src/qos.rs (hunk 23 above)
[the remark this answers is not in the document]
Comment (packgron) [reply crates/telio-nurse/src/qos.rs:527]: Damn, this is clearly a typo, fix it :sweat_smile:

**Review by packgron — APPROVED** [review]
+0.5
